### PR TITLE
Remove shim

### PR DESCRIPTION
**Pull request by sipsma**
This PR gets rid of our shim and migrates all of its functionality to run directly in the engine instead.

Original motivation: as part of [the general engine refactoring/cleanup](https://github.com/dagger/dagger/issues/6747#issuecomment-2104860737) I've been working through I realized we would be able to simplify a lot if we served nested clients (e.g. module functions and nested execs) directly from our executor. Among other things, it would allow us to be more "stateless" and not need to pre-register nested clients (which is a confusing and bug prone set of code paths).

But at that point it felt like the right time to just bite the bullet and rm the shim entirely, which is possible now thanks to our new worker/executor and has a ton of other side benefits, e.g.
* Better performance
   * starting a container only requires starting runc, not "outer shim"->runc->"inner shim"
   * service health checks and tunneling don't need to start a container at all anymore. They only need a raw network namespace, not a full container (https://github.com/dagger/dagger/pull/7367/files#diff-1f2692364baf2a0e130fb4ec0eefc7a170e52149b1beb5114dd4e8a9feec0dfcR68)
   * nested clients will be able to connect direct over http to the engine, no grpc tunneling required
   * I don't believe these are bottlenecks at our current stage in a lot of cases, but they become more prominent in use cases that result in lots of short execs being spun up and/or a ton of data flowing back and forth between modules+engine.
* More debuggability
   * right now when the shim has a problem, it can be very hard to debug; I often resort to returning special exit codes to figure out what's going wrong, which is about as low fidelity as it gets. Now it's all just in the engine process so we can just log as needed.
* Simplicity
   * don't need to smuggle information to the shim through special env vars (and worry about cache invalidation, etc.)
* Easier customization
   * switching to a non-runc runtime (like containerd) won't involve hacks to shove the shim into the equation somewhere
* Security
   * nested clients won't have access to the whole engine socket; they only have access to make calls in their session
* Various other smaller benefits like avoiding problems with overriding user exec init processes, etc.

---

Status: I've migrated everything now, just some cleanup and testing left:
- [x] double check ONLCR stuff by running vim in an interactive terminal with line numbers
   - still can't replicate any problems with this, so my best guess until proven otherwise is that the more recent change to use the CLI's raw terminal rather than going through an emulator invalidated the need for it? It makes sense on some level in that I don't see other tools built on top of the same containerd/runc stack, such as docker, needing to change `ONLCR` (afaik anyways) and those work interactively. cc @vito 
- [x] double check otel live exporting w/ bass
- [x] double check healthcheck/tunneling logs usage of GlobalLogger, see if output can look same as before for clients
   - improved here: https://github.com/dagger/dagger/pull/7367/commits/f3a48c7d9efd52795fab08aeac33bad35d378303 should look the same as before now
- [x] see whether our lack of an automatic init in each container is a meaningfully breaking change
   - I think this is a meaningful issue; I also noticed now that the `python -m http.server` service we use in tests no longer responds to `SIGTERM` (though it does respond to `SIGINT`). [This explains why](https://unix.stackexchange.com/a/735792) and how it's related to the fact that we now allow user processes to be init.
   - Going with packaging [dumb-init](https://github.com/Yelp/dumb-init) into the engine container and injecting it as the init for all containers for now in order to avoid this being a breaking change: https://github.com/dagger/dagger/pull/7367/commits/c99840a9f6e2dbbbbd827315978b1fdec315c7c8
   - This approach should avoid all the weird unexpected behavior of not having an init in the container while also:
      - being extremely lightweight; dumb-init is 20kb statically linked and extremely simple (hence the name)
      - not being strictly required; we can easily add an option to the container API to disable this for use cases that require the user exec is the init (e.g. testing `systemd`, etc.)
- [x] manually test dagger call with local dirs; I saw some unexpected error at one point that could be related to filesync tweaks but don't see this in integ tests so far, so worth a manual double check
   - fixed that here https://github.com/dagger/dagger/pull/7367/commits/130e79fcc85e0cc3f94e5f160de6f848c750abc4

I will of course also cleanup the commit history here with more thorough explanations of the changes.

### CODE REVIEW

File: engine/buildkit/executorSpec.go
Comment (jedevc) [inline engine/buildkit/executorSpec.go:1]: I like this approach, it makes the executor *significantly* easier to understand, good refactoring :tada:

Of course, now I'm imagining how these callbacks could operate on a `[T any]` type, and then potentially upstream into buildkit? There's not a lot of dagger-specific logic left in `executor.go`.

Maybe again I'm misreading it, but suddenly it looks a lot more doable than when we discussed it previously in https://github.com/dagger/dagger/pull/7228#discussion_r1586129145, now the hook mechanism here contains almost *all* of the complexity - and is *so*, *so*, much more readable.

---

For context - I promise I'm not trying to be super nitpicky here :sweat_smile: The reason I'm keen on not having a hard-fork of these files as much is because we've definitely had some horrible context-lifetime bugs, etc with runc in `callWithIO`. These were really nasty to track down and understand, and I'm just trying to come up with ways to avoid divergence for some of those code paths.

**Review by jedevc — COMMENTED** [review]
Ahhh this is neat :tada: Really nice to have the `shim` gone - even just from an architectural perspective, the practical performance implications are super great as well. Agreed on the potential containerd-worker stuff as well - since containerd *already* has it's own shim layer for calling runc, having another layer there was horrible.

Despite the large diff, it's pretty easy to read - most of the code is just moved, and the big refactorings are mostly fairly easy to understand. One big thing to call out though - we definitely need to make sure to test the new `Filesyncer` on Windows, there's a lot of fragile stuff there (ahhh we need tests for this #3462).

I'm not *as* familiar with the services/networking here, so not really able to comment on how that's changed - I like the `RunInNetNS` function though, that's neat.

> I will of course also cleanup the commit history here with more thorough explanations of the changes.

This would definitely be appreciated :heart: If you do this, merging without a squash would also be great, that way we can have a bit more insight with blame :tada:

File: core/schema/modulesource.go
Comment (sipsma) [reply core/schema/modulesource.go:1066]: Ah okay, it's not strictly necessary, I put it here because a bug I fixed would have failed much earlier and more comprehensibly with this check in place.

I will remove it for now unless I can find some way of implementing it cross-platform. It's tbh quite annoying that the Go stdlib makes the `filepath` implementation for other platforms completely inaccessible unless you compile the entire binary for that platform... I'm sure there's some good reason for it though 😄

File: engine/client/client.go
Comment (sipsma) [reply engine/client/client.go:159]: Yeah I am gonna clean those up; this one in particular I'll just remove I think. I suspect it's theoretically true that keepalives aren't needed when connecting direct over HTTP rather than doing grpc tunnelling (which is thankfully no longer needed for modules), but it's such a subtle change that could cause obscure race conditions that it's better to save for a different time I think.

File: engine/client/filesync.go
Comment (sipsma) [reply engine/client/filesync.go:37]: Right, did you end up with a good way of testing dagger on windows? I'll just fallback to running a VM on my laptop, but double checking there's not some easy way to run tests on a GHA windows runner manually or something.

File: engine/buildkit/executorSpec.go
Comment (sipsma) [reply engine/buildkit/executorSpec.go:1]: Totally, it's much more doable than before, though I think it might require refactoring upstream to use the same sort of approach. Honestly the simplest thing we could do upstream might just be to capitalize some of the structs/funcs it uses so we can import them and use them?

File: core/containerExec.go
Comment (sipsma) [reply core/containerExec.go:1]: Yeah I'll change this, but worth noting that I'm pretty sure this technically makes the file have the go build tag of `exec`. Completely harmless right now since we don't use go build tags, but theoretically may need to adjust down the line if we ever use those (afaik, I could be misremembering something).

File: core/schema/modulesource.go
Comment (sipsma) [reply core/schema/modulesource.go:1066]: At first glance, the implementation of `IsAbs` for windows is not too outrageous appearing: https://cs.opensource.google/go/go/+/refs/tags/go1.22.3:src/path/filepath/path_windows.go;l=59

So maybe we can just have our own util that does it? Requires the client tell us their platform, but that's extremely trivial plumbing.

File: engine/client/filesync.go
Comment (sipsma) [reply engine/client/filesync.go:37]: (I just installed it in a VM, it was less painful than I remember other than the fact that I can't run nested docker in there and have to connect to a remote docker on linux over tailscale 🤪)

File: core/containerExec.go
Comment (vito) [reply core/containerExec.go:1]: AFAIK that only applies to certain tags - `foo_bar.go` won't mean a `bar` build tag. According to ChatGPT the only magic tags in filenames are for each OS, architecture, and `_test.go`.

Comment (sipsma) [reply core/containerExec.go:1]: Ohhh nice, I didn't realize the filename convention only applied to os/arch: https://pkg.go.dev/cmd/go#hdr-Build_constraints Thanks, good to know!

**Review by vito — APPROVED** [review]
A mic drop of a PR if I've ever seen one :tada:

File: engine/client/filesync.go
Comment (sipsma) [reply engine/client/filesync.go]: Okay fixed this here: https://github.com/dagger/dagger/pull/7367/commits/48e774a1e499482d67ae8c05760663b951751c95

Tested on windows manually and could still call modules loaded from local sources, so I *think* we are good there.

File: core/schema/modulesource.go
Comment (sipsma) [reply core/schema/modulesource.go]: I just went with rm'ing the check for now, too much extra changes on top of an already large PR otherwise: https://github.com/dagger/dagger/commit/48e774a1e499482d67ae8c05760663b951751c95

File: core/c2h.go
Comment (sipsma) [reply core/c2h.go:32]: I went to do this and just saw `FrontendOrBackend` already exists haha: https://github.com/sipsma/dagger/blob/48e774a1e499482d67ae8c05760663b951751c95/core/net.go#L81-L81

Updating to that 😄

File: core/containerExec.go
Comment (sipsma) [reply core/containerExec.go:346]: Ah good catch, actually this should be entirely invalidated now since the metamount isn't even actually mounted into the container anymore.

We still need to "mount" it from the perspective of LLB in order for its cached result to be tied to the result of the actual exec, but we remove it from the OCI spec in the executor and just write to it directly from there. Kind of nicer in general since I suppose previously it was probably possible for user execs to try shenanigans w/ write access to it.

So I'll just remove this comment now.

**Comment by sipsma** [comment]
> This would definitely be appreciated ❤️ If you do this, merging without a squash would also be great, that way we can have a bit more insight with blame 🎉

Done! Did some consolidation to cut the number of commits in half and added more context to the messages as needed.

Hopefully I didn't fat finger anything while resolving conflicts during it. I'll double check with fresh eyes on Monday before pressing merge (will avoid the squash in this case too like you said).

File: engine/client/filesync.go
Comment (jedevc) [reply engine/client/filesync.go]: > end up with a good way of testing dagger on windows

Nope :( I just ended up using my personal windows install that I have around for gaming.

Will try and find another moment to give it a go :tada:

File: engine/buildkit/executorSpec.go
Comment (jedevc) [reply engine/buildkit/executorSpec.go:1]: > I think it might require refactoring upstream to use the same sort of approach

Yup, I think that's what I was imagining - the upstream implementation as is is quite complex, I think this approach is much easier to read and understand, so it definitely has my vote :)

**Comment by sipsma** [comment]
Double-checked the diff after doing commit cleanup, everything looks correct, so I'm gonna do rebase+merge now while it's still up to date with main to get the big diff out of the way for any other pending PRs w/ conflicts. cc @jedevc if you find anything off with Windows still let me know and will fixup in a follow-up!

**Pull request merged**
